modules/apps/document-library/document-library-service/src/main/java/com/liferay/document/library/internal/upgrade/v3_2_0/UpgradeDLFileEntryType.java

```diff
@@ -14,9 +14,221 @@
+import com.liferay.portal.kernel.util.PortalUtil;
+import com.liferay.portal.kernel.util.StringUtil;
+import com.liferay.portal.kernel.uuid.PortalUUIDUtil;
+
+import java.sql.Date;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+
 /**
  * @author Adolfo Pérez
  */
 public class UpgradeDLFileEntryType
 	extends com.liferay.portal.upgrade.v7_3_x.UpgradeDLFileEntryType {
+
+	@Override
+	protected void doUpgrade() throws Exception {
+		super.doUpgrade();
+
+		try (PreparedStatement ps1 = connection.prepareStatement(
+				StringBundler.concat(
+					"select companyId, fileEntryTypeId, fileEntryTypeKey, ",
+					"groupId, name, userId, userName from DLFileEntryType ",
+					"where (dataDefinitionId IS NULL OR dataDefinitionId = 0) ",
+					"and fileEntryTypeKey != 'BASIC-DOCUMENT'"));
+			PreparedStatement ps2 = AutoBatchPreparedStatementUtil.autoBatch(
+				connection.prepareStatement(
+					"update DLFileEntryType set dataDefinitionId = ? where " +
+						"fileEntryTypeId = ? "));
+			ResultSet rs = ps1.executeQuery()) {
+
+			String ddmStructureDefinition = StringUtil.read(
+				UpgradeDLFileEntryType.class.getResourceAsStream(
+					"dependencies/ddm-structure.json"));
+
+			String ddmStructureLayoutDefinition = StringUtil.read(
+				UpgradeDLFileEntryType.class.getResourceAsStream(
+					"dependencies/ddm-structure-layout.json"));
+
+			while (rs.next()) {
+				long ddmStructureId = _insertDDMStructure(
+					rs.getLong("groupId"), rs.getLong("companyId"),
+					rs.getLong("userId"), rs.getString("userName"),
+					rs.getString("name"), ddmStructureDefinition);
+
+				_insertDDMStructureLink(
+					rs.getLong("companyId"), rs.getLong("fileEntryTypeId"),
+					ddmStructureId);
+
+				long ddmStructureVersionId = _insertDDMStructureVersion(
+					rs.getLong("groupId"), rs.getLong("companyId"),
+					rs.getLong("userId"), rs.getString("userName"),
+					rs.getString("name"), ddmStructureId,
+					ddmStructureDefinition);
+
+				_insertDDMStructureLayout(
+					rs.getLong("groupId"), rs.getLong("companyId"),
+					rs.getLong("userId"), rs.getString("userName"),
+					rs.getString("name"), ddmStructureId, ddmStructureVersionId,
+					ddmStructureLayoutDefinition);
+
+				ResourceLocalServiceUtil.addResources(
+					rs.getLong("companyId"), rs.getLong("groupId"),
+					rs.getLong("userId"),
+					ResourceActionsUtil.getCompositeModelName(
+						DLFileEntryMetadata.class.getName(),
+						DDMStructure.class.getName()),
+					ddmStructureId, false, false, false);
+
+				ps2.setLong(1, ddmStructureId);
+
+				ps2.setLong(2, rs.getLong("fileEntryTypeId"));
+
+				ps2.addBatch();
+			}
+
+			ps2.executeBatch();
+		}
+	}
+
+	private long _insertDDMStructure(
+			long groupId, long companyId, long userId, String userName,
+			String name, String ddmStructureDefinition)
+		throws Exception {
+
+		try (PreparedStatement ps2 = connection.prepareStatement(
+				"INSERT INTO DDMStructure VALUES (0, 0, ?, ?, ?, ?, ?, ?, ?, " +
```

Comment:
To avoid portability issues, we need to specify the column names in the insert statements. Something like this:
```
insert into DDMStructure (structureId, groupId, companyId, ...) values (0, 0, ?, ...)
```